cpp-client/deephaven/client/src/client.cc

```diff
@@ -558,6 +591,425 @@ std::string TableHandle::toString(bool wantHeaders) const {
+      throw std::runtime_error(DEEPHAVEN_DEBUG_MSG(message));
+    }
+  }
+}
+
+grpcMathContext convertMathContext(MathContext mctx) {
+  int32_t precision;
+  RoundingMode roundingMode;
+  switch (mctx) {
+    case MathContext::UNLIMITED: {
+      precision = 0;
+      roundingMode = RoundingMode::MathContext_RoundingMode_HALF_UP;
+      break;
+    }
+    case MathContext::DECIMAL32: {
+      precision = 7;
+      roundingMode = RoundingMode::MathContext_RoundingMode_HALF_EVEN;
+      break;
+    }
+    case MathContext::DECIMAL64: {
+      precision = 16;
+      roundingMode = RoundingMode::MathContext_RoundingMode_HALF_EVEN;
+      break;
+    }
+    case MathContext::DECIMAL128: {
+      precision = 34;
+      roundingMode = RoundingMode::MathContext_RoundingMode_HALF_EVEN;
+      break;
+    }
+    default: {
+      auto message = stringf("Unexpected MathContext %o", (int)mctx);
+      throw std::runtime_error(DEEPHAVEN_DEBUG_MSG(message));
+    }
+  }
+  grpcMathContext result;
+  result.set_precision(precision);
+  result.set_rounding_mode(roundingMode);
+  return result;
+}
+
+UpdateByEmOptions convertOperationControl(const OperationControl &oc) {
+  auto onNull = convertBadDataBehavior(oc.onNull);
+  auto onNan = convertBadDataBehavior(oc.onNaN);
+  auto bigValueContext = convertMathContext(oc.bigValueContext);
+
+  UpdateByEmOptions result;
+  result.set_on_null_value(onNull);
+  result.set_on_nan_value(onNan);
+  *result.mutable_big_value_context() = std::move(bigValueContext);
+  return result;
+};
+
+/**
+ * decayTime will be specified as either std::chrono::nanoseconds, or as a string.
+ * If it is nanoseconds, we set the nanos field of the UpdateByWindowTime proto. Otherwise (if it is
+ * a string), then we set the duration_string field.
+ */
+UpdateByWindowTime convertDecayTime(std::string timestampCol, durationSpecifier_t decayTime) {
+  struct visitor_t {
+    void operator()(std::chrono::nanoseconds nanos) {
+      result.set_nanos(nanos.count());
+    }
+    void operator()(std::string duration) {
+      *result.mutable_duration_string() = std::move(duration);
+    }
+    UpdateByWindowTime result;
+  };
+  visitor_t v;
+  // Unconditionally set the column field with the value from timestampCol
+  *v.result.mutable_column() = std::move(timestampCol);
+
+  // Conditionally set either the nanos field or the duration_string with the nanoseconds or string
+  // part of the variant.
+  std::visit(v, std::move(decayTime));
+  return std::move(v.result);
+}
+
+class UpdateByBuilder {
+public:
+  explicit UpdateByBuilder(std::vector<std::string> cols) {
+    moveVectorData(std::move(cols), gup_.mutable_column()->mutable_match_pairs());
+  }
+
+  template<typename MEMBER>
+  void touchEmpty(MEMBER mutableMember) {
+    (void)(gup_.mutable_column()->mutable_spec()->*mutableMember)();
+  }
+
+  template<typename MEMBER>
+  void setNullBehavior(MEMBER mutableMember, DeltaControl deltaControl) {
```

Comment:
deltaContol can be const?
In general in several of the methods that follow, arguments that are not going to be `std::move`'d could be const.

Reply:
Maybe I should do a big sweep and apply const everywhere. That feels like that could be its own PR.

Reply:
No need.